superset/views/utils.py

```diff
@@ -131,8 +131,21 @@ def get_form_data(slice_id=None, use_slice_data=False):
         slc = db.session.query(models.Slice).filter_by(id=slice_id).one_or_none()
         if slc:
             slice_form_data = slc.form_data.copy()
+            # we want to get adhoc_filters from both form_data and slice_form_data
+            adhoc_filters = form_data.get("adhoc_filters", []) + slice_form_data.get("adhoc_filters", [])
             slice_form_data.update(form_data)
             form_data = slice_form_data
+            # ensure adhoc_filters are unique
```

Comment:
nitpick this logic seems to have a lot of heuristics built into it - sometimes it's nice to externalize into a function to make sure the danger doesn't get lost - for example "infer_unique_filters" or something.